core/src/cluster_info_repair_listener.rs

```diff
@@ -0,0 +1,683 @@
+use crate::blocktree::Blocktree;
+use crate::cluster_info::ClusterInfo;
+use crate::crds_value::EpochSlots;
+use crate::result::Result;
+use crate::service::Service;
+use byteorder::{ByteOrder, LittleEndian};
+use rand::seq::SliceRandom;
+use rand::SeedableRng;
+use rand_chacha::ChaChaRng;
+use solana_metrics::datapoint;
+use solana_runtime::epoch_schedule::EpochSchedule;
+use solana_sdk::pubkey::Pubkey;
+use std::cmp::min;
+use std::collections::HashMap;
+use std::mem;
+use std::net::SocketAddr;
+use std::net::UdpSocket;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, RwLock};
+use std::thread::{self, sleep, Builder, JoinHandle};
+use std::time::Duration;
+
+pub const REPAIRMEN_SLEEP_MILLIS: usize = 1000;
+pub const REPAIR_REDUNDANCY: usize = 3;
+pub const BLOB_SEND_SLEEP_MILLIS: usize = 2;
+pub const NUM_BUFFER_SLOTS: usize = 100;
+
+pub struct ClusterInfoRepairListener {
+    thread_hdls: Vec<JoinHandle<()>>,
+}
+
+impl ClusterInfoRepairListener {
+    pub fn new(
+        blocktree: &Arc<Blocktree>,
+        exit: &Arc<AtomicBool>,
+        cluster_info: Arc<RwLock<ClusterInfo>>,
+        epoch_schedule: EpochSchedule,
+    ) -> Self {
+        let exit = exit.clone();
+        let blocktree = blocktree.clone();
+        let thread = Builder::new()
+            .name("solana-cluster_info_repair_listener".to_string())
+            .spawn(move || {
+                // Maps a peer to the last timestamp and root they gossiped
+                let mut peer_roots: HashMap<Pubkey, (u64, u64)> = HashMap::new();
+                let _ = Self::recv_loop(
+                    &blocktree,
+                    &mut peer_roots,
+                    exit,
+                    &cluster_info,
+                    epoch_schedule,
+                );
+            })
+            .unwrap();
+        Self {
+            thread_hdls: vec![thread],
+        }
+    }
+
+    fn recv_loop(
+        blocktree: &Blocktree,
+        peer_roots: &mut HashMap<Pubkey, (u64, u64)>,
+        exit: Arc<AtomicBool>,
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        epoch_schedule: EpochSchedule,
+    ) -> Result<()> {
+        let socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+        let my_id = cluster_info.read().unwrap().id();
+        let mut my_gossiped_root = 0;
+
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                return Ok(());
+            }
+
+            let peers = cluster_info.read().unwrap().gossip_peers();
+            let mut peers_needing_repairs: HashMap<Pubkey, EpochSlots> = HashMap::new();
+
+            // Iterate through all the known nodes in the network, looking for ones that
+            // need repairs
+            for peer in peers {
+                let last_update_ts = Self::get_last_ts(peer.id, peer_roots);
+                let my_root =
+                    Self::update_my_gossiped_root(&my_id, cluster_info, &mut my_gossiped_root);
```

Comment:
is gossiping 0 as our root correct when a node starts from a ledger?

Reply:
should we update our gossiped root every time we look at a peer?

Reply:
After a node starts from ledger, it initializes its gossiped root based on the most up to date root in BankForks in RepairService. 

This is a poorly named function, `update_my_gossiped_root` actually just reads our root from gossip into a local variable.